include/nil/crypto3/multiprecision/cpp_int_modular.hpp

```diff
@@ -0,0 +1,807 @@
+                static constexpr const bool value = true;
+            };
+
+            //
+            // Now define the various data layouts that are possible.
+            // For modular we only use fixed precision (i.e. no allocator), unsigned type with limb-usage count:
+            //
+            template<unsigned Bits>
+            class cpp_int_modular_base<Bits, false> {
+            public:
+                using limb_pointer = limb_type*;
+                using const_limb_pointer = const limb_type*;
+
+                struct scoped_shared_storage {
+                    constexpr scoped_shared_storage(const cpp_int_modular_base&, unsigned) {
+                    }
+                    constexpr void deallocate(unsigned) {
+                    }
+                };
+                //
+                // Interface invariants:
+                //
+                static_assert(Bits > sizeof(double_limb_type) * CHAR_BIT,
+                              "Template parameter Bits is inconsistent with the parameter trivial - did you "
+                              "mistakingly try to override the trivial parameter?");
+
+                static constexpr unsigned limb_bits = sizeof(limb_type) * CHAR_BIT;
+                static constexpr limb_type max_limb_value = ~static_cast<limb_type>(0u);
+                static constexpr unsigned internal_limb_count =
+                    Bits / limb_bits + ((Bits % limb_bits) ? 1 : 0);
+                static constexpr limb_type upper_limb_mask =
+                    (Bits % limb_bits) ? (limb_type(1) << (Bits % limb_bits)) - 1 : (~limb_type(0));
+                static_assert(internal_limb_count >= 2,
+                              "A fixed precision integer type must have at least 2 limbs");
+
+            private:
+                union data_type {
+                    // m_data[0] contains the lowest bits.
+                    limb_type m_data[internal_limb_count];
+                    limb_type m_first_limb;
+                    double_limb_type m_double_first_limb;
+
+                    constexpr data_type() : m_data{0} { } // We are requred to set this to 0 to make it constexpr.
+
+                    constexpr data_type(const data_type& o) 
+                        : m_data{0} {
+#ifndef BOOST_MP_NO_CONSTEXPR_DETECTION
+                        if (BOOST_MP_IS_CONST_EVALUATED(internal_limb_count)) {
+                            for (unsigned i = 0; i < internal_limb_count; ++i)
+                                m_data[i] = o.m_data[i];
+                        } else {
+#endif
+                            if (this != &o) {
+                                std::memcpy(m_data, o.m_data, internal_limb_count * sizeof(limb_type));
+                            }
+                        }
+                    }
+
+                    constexpr data_type(limb_type i) : m_data {i} {
+                    }
+#ifndef BOOST_MP_NO_CONSTEXPR_DETECTION
```

Comment:
Looks like the only place this constructor is used is the constructor below:
```
constexpr data_type(double_limb_type i) : m_double_first_limb(i)
      {
#ifndef BOOST_MP_NO_CONSTEXPR_DETECTION
         if (BOOST_MP_IS_CONST_EVALUATED(m_double_first_limb))
         {
            data_type t(static_cast<limb_type>(i & max_limb_value), static_cast<limb_type>(i >> limb_bits));
            *this = t;
         }
#endif
```
And because this only call is in "#ifndef BOOST_MP_NO_CONSTEXPR_DETECTION", they have placed the upper constructor in the same #ifndef. We never all this constructor from the "outer code".